molecule/provisioners/libvirtprovisioner.py

```diff
@@ -0,0 +1,923 @@
+        targz = tarfile.open(boxfile, mode='r:gz')
+        members = targz.getmembers()
+        for member in members:
+            if member.name.endswith('.img'):
+                targz.extract(member, self._pool_path)
+                os.rename(
+                    os.path.join(self._pool_path, member.name),
+                    os.path.join(self._pool_path, image['name'] + '.img'))
+
+    def _create_volume(self, pool, instance):
+        """
+        Create a libvirt storage volume for an instance
+
+        :param: Libvirt Storage Pool object
+        :param: a dict that describes an instance
+        """
+        utilities.print_info("Creating libvirt volume for instance {}".format(
+            instance['name']))
+        vol = ET.Element('volume', type='file')
+        ET.SubElement(vol, 'name').text = instance['name'] + '.img'
+        ET.SubElement(vol, 'capacity', unit='GiB').text = '40'
+        target = ET.SubElement(vol, 'target')
+        ET.SubElement(target, 'format', type='qcow2')
+        permissions = ET.SubElement(target, 'permissions')
+        # TODO: FIXME: Glaring security hole for anyone who does not trust every user on the system running libvirt
+        ET.SubElement(permissions, 'mode').text = '0666'
+        backing = ET.SubElement(vol, 'backingStore')
+        ET.SubElement(backing, 'path').text = os.path.join(
+            self._pool_path, instance['image']['name'] + '.img')
+        ET.SubElement(backing, 'format', type='qcow2')
+        volxml = ET.tostring(vol)
+        LOG.debug("\tXMLDesc: {}".format(volxml))
+        newvol = pool.createXML(volxml)
+        utilities.print_success("\tCreated volume for {}.\n".format(instance[
+            'name']))
+
+    def _destroy_volume(self, pool, instance):
+        """
+        Destroy a libvirt storage volume for an instance
+
+        :param: Libvirt Storage Pool object
+        :param: a dict that describes an instance
+        """
+        utilities.print_info(
+            "\t\tDestroying libvirt volume for instance {}".format(instance[
+                'name']))
+        try:
+            vol = pool.storageVolLookupByName(instance['name'] + '.img')
+            vol.delete()
+        except libvirt.libvirtError:
+            LOG.warning("\t\tNo volume for {}".format(instance['name']))
+            return
+        utilities.print_success('\t\tDestroyed libvirt volume for {}'.format(
+            instance['name']))
+
+    def _populate_image(self, instance):
+        """
+        Get this instance's image config.
+
+        Any config specified for this instance is merged with (overriding) any config that already exists for an image of the same name.
+
+        :param: a dict of an instance
+        :return: the final image dict for that instance
+        """
+        image = instance.get(
+            'image', self.molecule.config.config['libvirt']['images'][0])
+        for k in ['source', 'ssh_user', 'ssh_key']:
+            image[k] = image.get(k, filter(
```

Comment:
This is massively inefficient because:
- you're doing the filtering again and again (in the loop), always returning the same value
- you're filtering *all* the elements and you only consider the first one

```
default_image = next(ref_image for ref_image in self.molecule.config.config['libvirt']['images'] if ref_image['name'] == image['name'])
for key in ['source', 'ssh_user', 'ssh_key']:
    image.setdefault(key, default_image[key])
```